src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -3240,25 +3241,50 @@ public Object stringToF(DynamicObject string, boolean strict,
         @Child private StringByteCharacterIndexNode byteIndexToCharIndexNode = StringNodesFactory.StringByteCharacterIndexNodeFactory.create(null);
         @Child private NormalizeIndexNode normalizeIndexNode = StringNodesFactory.NormalizeIndexNodeGen.create(null, null);
 
-        @Specialization(guards = { "isRubyString(pattern)", "isBrokenCodeRange(pattern)" })
-        public DynamicObject stringIndexBrokenCodeRange(DynamicObject string, DynamicObject pattern, int start) {
+        @Specialization(guards = { "isRubyString(pattern)", "isSingleByteSearch(string, pattern)" })
+        public Object stringIndex(VirtualFrame frame, DynamicObject string, DynamicObject pattern, int start,
+                                  @Cached("create()") RopeNodes.BytesNode bytesNode,
+                                  @Cached("createBinaryProfile()") ConditionProfile badStartProfile,
+                                  @Cached("create()") BranchProfile foundMatchProfile) {
+            final Rope sourceRope = rope(string);
+            final int end = sourceRope.byteLength();
+            final byte[] sourceBytes = bytesNode.execute(sourceRope);
+            final byte searchByte = bytesNode.execute(rope(pattern))[0];
+
+            if (badStartProfile.profile(start < 0 || start > end)) {
+                return nil();
+            }
+
+            for (int i = start; i < end; i++) {
+                if (sourceBytes[i] == searchByte) {
+                    foundMatchProfile.enter();
+                    return i;
+                }
+            }
+
```

Comment:
It would be good to add a branch profile here too so it's clear to Graal whether nil can be returned.